### PR TITLE
[Feature] Support correlated scalar non-agg subquery

### LINKED ISSUE #9280: [Feature] Support correlated scalar non-agg subquery
## Feature request

Support correlated scalar non-agg subquery of `SELECT/WHERE/HAVING` clause

For example

```sql
DROP TABLE IF EXISTS `t0`;
CREATE TABLE IF NOT EXISTS `t0` (
  `v1` int(11) NULL,
  `v2` int(11) NULL
) ENGINE=OLAP
DUPLICATE KEY(`v1`)
DISTRIBUTED BY HASH(`v1`) BUCKETS 10
PROPERTIES (
 "replication_num" = "1"
);

DROP TABLE IF EXISTS `t1`;
CREATE TABLE IF NOT EXISTS `t1` (
  `v1` int(11) NULL,
  `v2` int(11) NULL
) ENGINE=OLAP
DUPLICATE KEY(`v1`)
DISTRIBUTED BY HASH(`v1`) BUCKETS 10
PROPERTIES (
 "replication_num" = "1"
);

INSERT INTO `t0`(v1, v2) VALUES
(1, 1), (2, 2), (3, 3), (4, 4), (5, 5), (NULL, 6), (7, 7);

INSERT INTO `t1`(v1, v2) VALUES
(2, 1), (3, 4), (4, 3), (5, 5), (NULL, 6), (7, NULL), (8,8);


-- Q1, subquery on SELECT clause
SELECT * FROM t0
WHERE t0.v2 > (
      SELECT t1.v2 FROM t1
      WHERE t0.v1 = t1.v1
);

-- Q2, subquery on WHERE clause
SELECT t0.*, (
      SELECT t1.v2 FROM t1
      WHERE t0.v1 = t1.v1
) FROM t0;

-- Q3, subquery on HAVING clause
SELECT v1, SUM(v2) FROM t0
GROUP BY v1
HAVING SUM(v2) > (
      SELECT t1.v2 FROM t1
      WHERE t0.v1 = t1.v1
);
```

I will illustrate the transform rule by the example query Q1, we can transfrom Q1 to Q1.1, as follows:

* Add an inline view `t1a`, which contains a group by clause, using inner column `t1.v1` of correlated predicate, i.e. `t0.v1 = t1.v1` as group by column
    * output `v1_Rows` used in `assert_true`
    * output `v2_Any` as the original output of subquery
    * output `t1.v1` used as join key
* Add a LOJ to produce the final output
* Rewrite the original predicate `t0.v2 > <subquery's output>` to `t0.v2 > t1a.v2_Any`
* Add a assert predicate `assert_true`

```sql
-- Q1.1
SELECT t0.*,
FROM t0 LEFT OUTER JOIN
(
    SELECT t1.v1, count(1) as v1_Rows, any_value(t1.v2) as v2_Any
    FROM t1 
    GROUP BY t1.v1
) as t1a
ON t0.v1 = t1a.v1
WHERE t0.v2 > t1a.v2_Any;
AND assert_true(t1a.v1_Rows <= 1);
```

### PR DESCRIPTION
**Pull request by liuyehcf**
## What type of PR is this：
- [ ] bug
- [x] feature
- [ ] enhancement
- [ ] refactor
- [ ] others

## Which issues of this PR fixes ：

Fixes #9280

## Problem Summary(Required) ：

Please refer to the issue(#9280) for details.

Here are the change lists of the pr:

1. Add correlated scalar non-agg transform logic to the exist rule `ScalarApply2JoinRule`
2. Add and implement built-in function `assert_true` to support the assertion.
3. Add ut

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/scalar/FoldConstantsRule.java [hunk 1]
```diff
@@ -23,7 +24,8 @@ public class FoldConstantsRule extends BottomUpScalarOperatorRewriteRule {
 
     @Override
     public ScalarOperator visitCall(CallOperator call, ScalarOperatorRewriteContext context) {
-        if (call.isAggregate() || notAllConstant(call.getChildren())) {
+        if (call.isAggregate() || notAllConstant(call.getChildren()) ||
+                FunctionSet.ASSERT_TRUE.equals(call.getFnName())) {
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/scalar/FoldConstantsRule.java:28]: Should handle in ScalarOperatorEvaluator evaluation method.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 57 / 57 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/ScalarApply2JoinRule.java|55|55|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rewrite/scalar/FoldConstantsRule.java|2|2|100.00%|[]|

File: be/src/exprs/vectorized/utility_functions.cpp [hunk 2]
```diff
@@ -2,6 +2,8 @@
 
 #include "exprs/vectorized/utility_functions.h"
 
+#include <simd/simd.h>
```
Comment (Youngwb) [inline be/src/exprs/vectorized/utility_functions.cpp:5]: This include is necessary？

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/ScalarApply2JoinRule.java [hunk 3]
```diff
@@ -48,32 +64,139 @@ public List<OptExpression> transform(OptExpression input, OptimizerContext conte
...
-            output.put(apply.getOutput(), apply.getSubqueryOperator());
+        // count aggregate
+        Map<ColumnRefOperator, CallOperator> aggregates = Maps.newHashMap();
+        CallOperator countRowsCallOp = SubqueryUtils.createCountRowsOperator();
+        ColumnRefOperator countRows =
+                factory.create("countRows", countRowsCallOp.getType(), countRowsCallOp.isNullable());
+        aggregates.put(countRows, countRowsCallOp);
+
+        // any_value aggregate
+        ScalarOperator subqueryOperator = apply.getSubqueryOperator();
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/ScalarApply2JoinRule.java:126]: Would better put all create concrete aggregate function call to a util class, then we could reuse some codes.

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/ScalarApply2JoinRule.java [hunk 4]
```diff
@@ -48,32 +64,139 @@ public List<OptExpression> transform(OptExpression input, OptimizerContext conte
...
+     * Transfer to:
+     *   SELECT t0.*
+     *   FROM t0 LEFT OUTER JOIN
+     *   (
+     *       SELECT t1.v1, count(1) as countRows, any_value(t1.v2) as anyValue
+     *       FROM t1
+     *       GROUP BY t1.v1
+     *   ) as t1a
+     *   ON t0.v1 = t1a.v1
+     *   WHERE t0.v2 > t1a.anyValue;
```
Comment (Youngwb) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/ScalarApply2JoinRule.java:93]: miss countRows?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/ScalarApply2JoinRule.java [hunk 5]
```diff
@@ -48,32 +64,139 @@ public List<OptExpression> transform(OptExpression input, OptimizerContext conte
...
+        Function assertTrueFn = Expr.getBuiltinFunction(FunctionSet.ASSERT_TRUE, new Type[] {Type.BOOLEAN},
+                Function.CompareMode.IS_IDENTICAL);
+        CallOperator assertTrueCallOp =
+                new CallOperator(FunctionSet.ASSERT_TRUE, Type.BOOLEAN,
+                        Collections.singletonList(countRowsPredicate),
+                        assertTrueFn);
+        LogicalJoinOperator joinOp = new LogicalJoinOperator.Builder()
+                .setJoinType(JoinOperator.LEFT_OUTER_JOIN)
+                .setOnPredicate(Utils.compoundAnd(correlationPredicatePair.first))
+                .setPredicate(assertTrueCallOp)
```
Comment (Youngwb) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/ScalarApply2JoinRule.java:165]: predicate t0.v2 > t1a.anyValue do not used in join operator? where this predicate used?

File: be/src/exprs/vectorized/utility_functions.cpp (hunk 2 above)
Comment (liuyehcf) [reply be/src/exprs/vectorized/utility_functions.cpp:5]: Not necessary, I'll remove it

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/scalar/FoldConstantsRule.java (hunk 1 above)
Comment (liuyehcf) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/scalar/FoldConstantsRule.java:28]: Ok

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/ScalarApply2JoinRule.java (hunk 5 above)
Comment (liuyehcf) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/ScalarApply2JoinRule.java:163]: This predicate is not belongs to this rule pattern, we just add a project operator to map `subquery output` to `anyValue`.

File: fe/fe-core/src/test/java/com/starrocks/sql/plan/FunctionTest.java [hunk 6]
```diff
@@ -0,0 +1,26 @@
+// This file is licensed under the Elastic License 2.0. Copyright 2021-present, StarRocks Limited.
+
+package com.starrocks.sql.plan;
+
+import org.junit.Test;
+
+public class FunctionTest extends PlanTestBase {
+    @Test
+    public void testAssertTrue() throws Exception {
```
[the remark this answers is not in the document]
Comment (liuyehcf) [reply fe/fe-core/src/test/java/com/starrocks/sql/plan/FunctionTest.java]: done

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/ScalarApply2JoinRule.java (hunk 4 above)
Comment (liuyehcf) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/ScalarApply2JoinRule.java]: done

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/ScalarApply2JoinRule.java (hunk 3 above)
Comment (liuyehcf) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/ScalarApply2JoinRule.java:127]: done

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 57 / 57 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/ScalarApply2JoinRule.java|53|53|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rewrite/scalar/FoldConstantsRule.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/SubqueryUtils.java|2|2|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/scalar/FoldConstantsRule.java (hunk 1 above)
Comment (liuyehcf) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/scalar/FoldConstantsRule.java]: done

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 57 / 57 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/ScalarApply2JoinRule.java|53|53|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rewrite/ScalarOperatorEvaluator.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/SubqueryUtils.java|2|2|100.00%|[]|

**Comment by liuyehcf** [comment]
> I have a question, This is the sql after tranform:
> 
> ```
> SELECT t0.*,
> FROM t0 LEFT OUTER JOIN
> (
>     SELECT t1.v1, count(1) as v1_Rows, any_value(t1.v2) as v2_Any
>     FROM t1 
>     GROUP BY t1.v1
> ) as t1a
> ON t0.v1 = t1a.v1
> WHERE t0.v2 > t1a.v2_Any
> AND assert_true(t1a.v1_Rows <= 1);
> ```
> 
> If the t1.v2 has two rows when t1 group by v1 and v1 =1 or some other value, one of t1.v2 is small than t0.v2(predicate t0.v2 > t1a.v2_Any is True), the other is bigger than t0.v2(predicate t0.v2 > t1a.v2_Any is False) Would it happen different result if multi execute, may be one is assert false. one is empty?

* For rows satisfied the predicate `t0.v2 > t1a.v2_Any`, if `t1a.v1_Rows > 1`, then error will be throwed during execution.
* For rows not satisfied the predicate `t0.v2 > t1a.v2_Any`, if `t1a.v1_Rows > 1`, the sql will be executed normally.

**Comment by liuyehcf** [comment]
> I have a question, This is the sql after tranform:
> 
> ```
> SELECT t0.*,
> FROM t0 LEFT OUTER JOIN
> (
>     SELECT t1.v1, count(1) as v1_Rows, any_value(t1.v2) as v2_Any
>     FROM t1 
>     GROUP BY t1.v1
> ) as t1a
> ON t0.v1 = t1a.v1
> WHERE t0.v2 > t1a.v2_Any
> AND assert_true(t1a.v1_Rows <= 1);
> ```
> 
> If the t1.v2 has two rows when t1 group by v1 and v1 =1 or some other value, one of t1.v2 is small than t0.v2(predicate t0.v2 > t1a.v2_Any is True), the other is bigger than t0.v2(predicate t0.v2 > t1a.v2_Any is False) Would it happen different result if multi execute, may be one is assert false. one is empty?

An error will be throwed during execution. if `t1.v2(1)` and `t1.v2(3)` straddle t0.v2(2), these two rows share the same `t1a.v1_Rows=2`, although `t1.v2(3)` is filtered by the predicate `t0.v2 > t1a.v2_Any`

**Review by kangkaisen — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 92 / 93 (98.92%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/OperatorBuilderFactory.java|2|3|66.67%|[85]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/ScalarApply2JoinRule.java|63|63|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PushDownPredicateProjectRule.java|9|9|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rewrite/ScalarOperatorEvaluator.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PruneProjectColumnsRule.java|5|5|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/MergeTwoProjectRule.java|9|9|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/SubqueryUtils.java|2|2|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/MergeTwoProjectRule.java [hunk 7]
```diff
@@ -4,25 +4,43 @@
...
+import java.util.Optional;
 
 public class MergeTwoProjectRule extends TransformationRule {
     public MergeTwoProjectRule() {
         super(RuleType.TF_MERGE_TWO_PROJECT, Pattern.create(OperatorType.LOGICAL_PROJECT)
                 .addChildren(Pattern.create(OperatorType.LOGICAL_PROJECT, OperatorType.PATTERN_LEAF)));
     }
 
+    @Override
+    public boolean check(OptExpression input, OptimizerContext context) {
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/MergeTwoProjectRule.java:30]: merge and add assert method to the first project plz

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/ScalarApply2JoinRule.java [hunk 8]
```diff
@@ -48,32 +64,145 @@ public List<OptExpression> transform(OptExpression input, OptimizerContext conte
...
+                factory.create("subquery_assertion", assertTrueCallOp.getType(), assertTrueCallOp.isNullable());
+        projectMap.put(assertion, assertTrueCallOp);
+        OptExpression assertProjectOpt =
+                OptExpression.create(new LogicalProjectOperator(projectMap), newLeftOuterJoinOpt);
+
+        // Step4: prune the assertion column
+        Map<ColumnRefOperator, ScalarOperator> prunedProjectMap = Maps.newHashMap(projectMap);
+        prunedProjectMap.remove(assertion);
+        OptExpression finalProjectOpt =
+                OptExpression.create(new LogicalProjectOperator(prunedProjectMap), assertProjectOpt);
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/ScalarApply2JoinRule.java:179]: seems useless

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/MergeTwoProjectRule.java (hunk 7 above)
Comment (liuyehcf) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/MergeTwoProjectRule.java]: Thanks, good suggestion! 
Done.

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/ScalarApply2JoinRule.java (hunk 8 above)
Comment (liuyehcf) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/ScalarApply2JoinRule.java]: Done.

**Comment by liuyehcf** [comment]
Sincerely thanks every single comment here, really helped me.

**Comment by liuyehcf** [comment]
run starrocks_fe_unittest

**Review by Seaven — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 83 / 86 (96.51%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/OperatorBuilderFactory.java|0|3|00.00%|[82, 83, 85]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/ScalarApply2JoinRule.java|59|59|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PushDownPredicateProjectRule.java|9|9|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rewrite/ScalarOperatorEvaluator.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PruneProjectColumnsRule.java|5|5|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/MergeTwoProjectRule.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/SubqueryUtils.java|2|2|100.00%|[]|

**Comment by liuyehcf** [comment]
run starrocks_admit_test

**Review by Youngwb — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Pull request merged**